data/prepare_dataset_giant_midi.py

```diff
@@ -0,0 +1,122 @@
+import os
+import math
+
+import fortepyan as ff
+from tqdm import tqdm
+from datasets import Value, Dataset, Features, Sequence, DatasetDict, load_dataset
+
+from data.quantizer import MidiQuantizer
+
+
+def process_record(piece: ff.MidiPiece, sequence_len: int, quantizer: MidiQuantizer) -> list[dict]:
+    piece_quantized = quantizer.quantize_piece(piece)
+
+    midi_filename = piece_quantized.source["midi_filename"]
+
+    piece.df["next_start"] = piece.df.start.shift(-1)
+    piece.df["dstart"] = piece.df.next_start - piece.df.start
+    piece.df["dstart"].fillna(0, inplace=True)
+
+    record = []
+
+    for subset, subset_quantized in zip(
+        piece.df.rolling(window=sequence_len, step=sequence_len),
+        piece_quantized.df.rolling(window=sequence_len, step=sequence_len),
+    ):
+        # rolling sometimes creates subsets with shorter sequence length, they are filtered here
+        if len(subset) != sequence_len:
+            continue
+
+        sequence = {
+            "midi_filename": midi_filename,
+            "pitch": subset.pitch.astype("int16").values.T,
+            "dstart": subset.dstart.astype("float32"),
+            "duration": subset.duration.astype("float32"),
+            "velocity": subset.velocity.astype("int16"),
+            "dstart_bin": subset_quantized.dstart_bin.astype("int8").values.T,
+            "duration_bin": subset_quantized.duration_bin.astype("int8").values.T,
+            "velocity_bin": subset_quantized.velocity_bin.astype("int8").values.T,
+        }
+
+        record.append(sequence)
+
+    return record
+
+
+def split_in_two(filenames: list[str], split_ratio: float = 0.8) -> tuple[list[str], list[str]]:
+    ds_length = len(filenames)
+
+    split = math.ceil(split_ratio * ds_length)
+
+    return filenames[:split], filenames[split:]
+
+
+def main():
+    # get huggingface token from environment variables
+    token = os.environ["HUGGINGFACE_TOKEN"]
+
+    # hyperparameters
+    sequence_len = 128
+
+    hf_dataset_path = "roszcz/giant-midi-sustain"
+    dataset = load_dataset(hf_dataset_path, split="train")
+
+    quantizer = MidiQuantizer(
+        n_dstart_bins=7,
+        n_duration_bins=7,
+        n_velocity_bins=7,
+    )
```

Comment:
Can you test the dataset speed without caching the quantization? 

It's problematic to cache datasets for all the combinations of bins we might want to try - it would be much better to do the quantization during train time - I think it should be fast enough, but may require some modification of the Quantizer class, so it doesn't operate on data frames, but something faster 🤔